### PR TITLE
Add support for RTSP Opus

### PR DESCRIPTION
**Pull request by ManishaJajoo**
Added Opus RTP packet reader and added support for Opus
playback through RTSP

Change-Id: Ib6702bd8aafd0bd782e89127ab907061ff06ccb3

### CODE REVIEW

File: libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpOpusReader.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+      ParsableByteArray data, long timestamp, int sequenceNumber, boolean rtpMarker) {
+    checkStateNotNull(trackOutput);
+
+    /* RFC7845 Section 3
+     * +---------+ +----------------+ +--------------------+ +-----
+     * |ID Header| | Comment Header | |Audio Data Packet 1 | | ...
+     * +---------+ +----------------+ +--------------------+ +-----
+     */
+    if (!foundOpusIDHeader) {
+      int currPosition = data.getPosition();
```
Comment (claincly) [inline libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpOpusReader.java:84]: What about maintaining the position in `isOpusIDHeader`? As if calling `isOpusIDHeader` does not have side effect on data position

File: libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpOpusReader.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+      checkArgument(isOpusIDHeader(data), "ID Header missing");
+
+      data.setPosition(currPosition);
+      List<byte[]> initializationData = OpusUtil.buildInitializationData(data.getData());
+      Format.Builder formatBuilder = payloadFormat.format.buildUpon();
+      formatBuilder.setInitializationData(initializationData);
+      trackOutput.format(formatBuilder.build());
+      foundOpusIDHeader = true;
+    } else if (!foundOpusCommentHeader) {
```
Comment (claincly) [inline libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpOpusReader.java:93]: What about checking ID header and comment header in one go? given they should be together in the packet.

File: libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpOpusReader.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+                "Received RTP packet with unexpected sequence number. Expected: %d; received: %d.",
+                expectedSequenceNumber, sequenceNumber));
+      }
+
+      // sending opus data
+      int size = data.bytesLeft();
+      trackOutput.sampleData(data, size);
+      long timeUs =
+          toSampleTimeUs(startTimeOffsetUs, timestamp, firstReceivedTimestamp, sampleRate);
+      trackOutput.sampleMetadata(timeUs, C.BUFFER_FLAG_KEY_FRAME, size, 0, null);
```
Comment (claincly) [inline libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpOpusReader.java:114]: add inline param name, like /* crytoData*/ = null

**Review by claincly — CHANGES_REQUESTED** [review]

File: libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/RtspMediaTrack.java [hunk 4]
```diff
@@ -121,6 +121,14 @@ public int hashCode() {
         checkArgument(!fmtpParameters.isEmpty());
         processAacFmtpAttribute(formatBuilder, fmtpParameters, channelCount, clockRate);
         break;
+      case MimeTypes.AUDIO_OPUS:
+        // RFC7587 Section 7
+        checkArgument(channelCount == 2, "Invalid channel count");
+        // RFC7587 Section 6.1
+        // the RTP timestamp is incremented with a 48000 Hz clock rate
+        // for all modes of Opus and all sampling rates.
+        checkArgument(clockRate == 48000, "Invalid sampling rate");
```
[the remark this answers is not in the document]
Comment (ManishaJajoo) [reply libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/RtspMediaTrack.java:130]: RtpOpusReader is not visible to RtspMediaTrack as the readers belong to a different package and are default
Added the constant to this class itself

File: libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpOpusReader.java (hunk 2 above)
Comment (ManishaJajoo) [reply libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpOpusReader.java:93]: We haven't been able to generate a streaming data with both headers being a part of same packet. Can you suggest something for validating the corresponding code changes?

File: libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpOpusReader.java [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    firstReceivedTimestamp = nextRtpTimestamp;
+    startTimeOffsetUs = timeUs;
+  }
+
+  // Internal methods.
+
+  private static boolean isOpusIDHeader(ParsableByteArray data) {
+    int sampleSize = data.limit();
+    String header = data.readString(8);
+    // Identification header RFC7845 Section 5.1
```
[the remark this answers is not in the document]
Comment (ManishaJajoo) [reply libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpOpusReader.java]: Can you please elaborate?
Are you suggesting to leave a javadoc explaining the connection between Opus and NAT Behavioral Requirements?

**Comment by claincly** [comment]
Merged internally. The issue will be marked as merged automatically when we do a push. Thanks for the contribution!

**Comment by claincly** [comment]
This branch is already merged internally. Can you submit the test as a separate PR? Thanks

File: libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpOpusReader.java [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+  public void consume(
+      ParsableByteArray data, long timestamp, int sequenceNumber, boolean rtpMarker) {
+    checkStateNotNull(trackOutput);
+
+    /* RFC7845 Section 3.
+     * +---------+ +----------------+ +--------------------+ +-----
+     * |ID Header| | Comment Header | |Audio Data Packet 1 | | ...
+     * +---------+ +----------------+ +--------------------+ +-----
+     */
+    if (!foundOpusIDHeader) {
```
Comment (claincly) [inline libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpOpusReader.java:82]: Sorry for not getting back to this sooner, I thought we have merged it.

All the logic here to find Opus ID and comment headers imply that the header and comment header is ONE PACKET long. But that is not what is specified in the RFC.

According to the RFC, the ID header is one PAGE long, and the comment header can span across many pages. And one page can span multiple RTP packets. Also, "Audio data packets might span page boundaries.  The first audio data page could have the 'continued packet' flag set"

Although this is working right now, it's not semantically correct. So please try to make it compliant with the RFC by supporting multiple packet pages. There are examples of one access unit spans multiple RTP packets, like in the H264/265 reader. It's probably hard to test, but given you have some tests going on, please add some test based on your interpretation to the RFC.

Comment (claincly) [inline libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpOpusReader.java:82]: Sorry for not getting back to this sooner, I thought we have merged it.

All the logic here to find Opus ID and comment headers imply that the header and comment header is ONE PACKET long. But that is not what is specified in the RFC.

According to the RFC, the ID header is one PAGE long, and the comment header can span across many pages. And one page can span multiple RTP packets. Also, "Audio data packets might span page boundaries.  The first audio data page could have the 'continued packet' flag set"

Although this is working right now, it's not semantically correct. So please try to make it compliant with the RFC by supporting multiple packet pages. There are examples of one access unit spans multiple RTP packets, like in the H264/265 reader. It's probably hard to test, but given you have some tests going on, please add some test based on your interpretation to the RFC.

**Pull request merged**